mackerel-plugin-openldap/lib/openldap.go

```diff
@@ -0,0 +1,355 @@
+				{Name: "statistics_referrals_monitorCounter", Label: "referrals", Diff: true},
+			},
+		},
+		"connections": {
+			Label: (labelPrefix + " connections"),
+			Unit:  "integer",
+			Metrics: []mp.Metrics{
+				{Name: "connections_total_monitorCounter", Label: "total connection", Diff: true},
+				{Name: "connections_current_monitorCounter", Label: "current connection", Diff: true},
+			},
+		},
+	}
+	if m.ReplBase != "" {
+		graphs["replications"] = mp.Graphs{
+			Label: (labelPrefix + " replication delay"),
+			Unit:  "integer",
+			Metrics: []mp.Metrics{
+				{Name: "replication_delay", Label: "replication delay sec", Diff: false},
+			},
+		}
+	}
+	return graphs
+}
+
+// Do the plugin
+func Do() {
+	optHost := flag.String("host", "localhost", "Hostname")
+	optPort := flag.String("port", "389", "Port")
+	optTLS := flag.Bool("tls", false, "TLS(ldaps)")
+	optInsecureSkipVerify := flag.Bool("insecureSkipVerify", false, "TLS accepts any certificate.")
+	optReplBase := flag.String("replBase", "", "replication base dn")
+	optReplMasterBind := flag.String("replMasterBind", "", "replication master bind dn")
+	optReplMasterHost := flag.String("replMasterHost", "", "replication master hostname")
+	optReplMasterTLS := flag.Bool("replMasterTLS", false, "replication master TLS(ldaps)")
+	optReplMasterPort := flag.String("replMasterPort", "389", "replication master port")
+	optReplMasterPass := flag.String("replMasterPW", "", "replication master bind password")
+	optReplLocalBind := flag.String("replLocalBind", "", "replicationlocalmaster bind dn")
+	optReplLocalPass := flag.String("replLocalPW", "", "replication local bind password")
+	optBindDn := flag.String("bind", "", "bind dn")
+	optBindPasswd := flag.String("pw", "", "bind password")
+	optTempfile := flag.String("tempfile", "", "Temp file name")
+	optPrefix := flag.String("metric-key-prefix", "openldap", "Metric key prefix")
+	flag.Parse()
+
+	var m OpenLDAPPlugin
+	m.TargetHost = fmt.Sprintf("%s:%s", *optHost, *optPort)
+	m.ReplMasterHost = fmt.Sprintf("%s:%s", *optReplMasterHost, *optReplMasterPort)
+	m.UseTLS = *optTLS
+	m.InsecureSkipVerify = *optInsecureSkipVerify
+	m.ReplBase = *optReplBase
+	m.ReplMasterUseTLS = *optReplMasterTLS
+	m.ReplMasterBind = *optReplMasterBind
+	m.ReplMasterPass = *optReplMasterPass
+	m.ReplLocalBind = *optReplLocalBind
+	m.ReplLocalPass = *optReplLocalPass
+	if m.InsecureSkipVerify {
+		m.UseTLS = true
+	}
+	m.BindDn = *optBindDn
+	m.BindPasswd = *optBindPasswd
+	m.Prefix = *optPrefix
+
+	if *optBindDn == "" {
+		logger.Errorf("bind is required.")
+		flag.PrintDefaults()
+		os.Exit(1)
+	}
+	var err error
+	if m.UseTLS {
+		m.l, err = ldap.DialTLS("tcp", m.TargetHost, &tls.Config{InsecureSkipVerify: m.InsecureSkipVerify})
+	} else {
+		m.l, err = ldap.Dial("tcp", m.TargetHost)
+	}
+	if err != nil {
+		logger.Errorf("Failed to Dial %s, err: %s", m.TargetHost, err)
+		os.Exit(1)
+	}
+	err = m.l.Bind(m.BindDn, m.BindPasswd)
```

Comment:
With this implementation, users cannot perform unauthenticated request to ldap.
I know in many cases users restrict ldap access and require binding, but it *might* be optional. (Yes, just *might*.)
What do you think?